### PR TITLE
feat: override `__sequence` on creating SST to save space and CPU

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

The sequence is `u64` and is allocated with write request. They are different from row to row and thus has bad compression efficiency in both CPU and space. We can assign the greatest sequence number to the file to reduce this overhead while keeping this mechanism works.

In TSBS test scenario with 100M rows, it can save 48MB of disk space. And in the metric monitor scenario, it can save up to 50%~60% in file size. We can also observe an (unstable) flush performance improvement.

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [x] API changes are backward compatible.
- [x] Schema or data changes are backward compatible.

### CODE REVIEW

**Comment by MichaelScofield** [comment]
good for merging ssts when ingesting them, better if the `__sequence` can be removed

**Comment by evenyag** [comment]
> good for merging ssts when ingesting them, better if the `__sequence` can be removed

We can share the same sequence because here we assume we always dedup all rows in the SST in non append-only mode.

File: src/mito2/src/sst/parquet/writer.rs [hunk 1]
```diff
@@ -112,9 +113,11 @@ where
     pub async fn write_all(
         &mut self,
         mut source: Source,
+        override_sequence: Option<SequenceNumber>, // override the `sequence` field from `Source`
         opts: &WriteOptions,
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/writer.rs:117]: Can we move the `override_sequence` to `WriteOptions`. It should be an option to control the writer.

File: src/mito2/src/memtable/time_series.rs [hunk 2]
```diff
@@ -199,6 +201,10 @@ impl Memtable for TimeSeriesMemtable {
         // so that we can ensure writing to memtable will succeed.
         self.update_stats(local_stats);
 
+        // update max_sequence
+        let sequence = kvs.mutation.sequence;
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:205]: The sequence in the `kvs` is the start sequence in the mutation. We can encapsulate a method to get the maximum sequence.
https://github.com/GreptimeTeam/greptimedb/blob/11bab0c47cce4fe09cd37763f7e1877d11178c7b/src/mito2/src/memtable/key_values.rs#L95-L104

We can also add a max sequence field to the metrics and update the sequence in `update_stats()`.

File: src/mito2/src/memtable/partition_tree.rs [hunk 3]
```diff
@@ -140,6 +141,12 @@ impl Memtable for PartitionTreeMemtable {
 
         self.update_stats(&metrics);
 
+        // update max_sequence
+        if res.is_ok() {
+            let sequence = kvs.mutation.sequence;
```
Comment (evenyag) [inline src/mito2/src/memtable/partition_tree.rs:146]: This is the start sequence of the mutation.

File: src/mito2/src/engine/basic_test.rs [hunk 4]
```diff
@@ -580,7 +580,7 @@ async fn test_region_usage() {
     flush_region(&engine, region_id, None).await;
 
     let region_stat = region.region_statistic();
-    assert_eq!(region_stat.sst_size, 2790);
+    assert_eq!(region_stat.sst_size, 2773);
```
Comment (evenyag) [inline src/mito2/src/engine/basic_test.rs:583]: Maybe we can assert the size > 0.

File: src/mito2/src/flush.rs [hunk 5]
```diff
@@ -337,6 +337,18 @@ impl RegionFlushTask {
...
+        let max_sequence = memtables
+            .iter()
+            .filter_map(|m| {
+                if !m.is_empty() {
+                    Some(m.stats().max_sequence())
+                } else {
+                    None
+                }
+            })
+            .max();
```
Comment (evenyag) [inline src/mito2/src/flush.rs:350]: I prefer not to share the max sequence across memtables.

File: src/mito2/src/flush.rs [hunk 6]
```diff
@@ -382,6 +395,7 @@ impl RegionFlushTask {
                 index_file_size: sst_info.index_metadata.file_size,
                 num_rows: sst_info.num_rows as u64,
                 num_row_groups: sst_info.num_row_groups,
+                max_sequence: max_sequence.unwrap_or(0),
```
Comment (evenyag) [inline src/mito2/src/flush.rs:398]: We can get the max sequence from this memtable.
```suggestion
                max_sequence: max_sequence.unwrap_or(0),
```

File: src/mito2/src/sst/file.rs [hunk 7]
```diff
@@ -134,6 +134,8 @@ pub struct FileMeta {
     /// the default value `0` doesn't means the file doesn't contains any rows,
     /// but instead means the number of rows is unknown.
     pub num_row_groups: u64,
+    /// Max sequence in this file
+    pub max_sequence: u64,
```
Comment (evenyag) [inline src/mito2/src/sst/file.rs:138]: Maybe we should rename it to `sequence`? It is a global sequence for all rows in the file.

TODO: While ingesting SSTs, we can also set this field to assign a sequence to the whole SST. We can do it in the mito engine so the mito engine can correctly bump the sequence number and assign it to the file meta. @MichaelScofield

Comment (evenyag) [inline src/mito2/src/sst/file.rs:138]: idea: Can we use `Option<NonZeroU64>`? So we won't forget that this field can be zero.

File: src/mito2/src/engine/basic_test.rs [hunk 8]
```diff
@@ -580,7 +580,7 @@ async fn test_region_usage() {
     flush_region(&engine, region_id, None).await;
 
     let region_stat = region.region_statistic();
-    assert_eq!(region_stat.sst_size, 2790);
+    assert!(region_stat.sst_size > 0); // Chief says this assert can ensure the size is counted.
```
Comment (evenyag) [inline src/mito2/src/engine/basic_test.rs:583]: 😕

**Review by evenyag — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**